File: apps/dsa-web/src/api/plugins.ts
// Copyright (c) 2026 SiinXu / StockPulse contributors
// SPDX-License-Identifier: AGPL-3.0-only
import { z } from 'zod';
import type { components } from '../types/api.generated';
import apiClient, { locallyRecoverableResourceConfig } from './index';
import { parseCamelCasePayload } from './parseCamelCasePayload';

type OpenApiPluginInfo = components['schemas']['PluginInfo'];
type OpenApiPluginListResponse = components['schemas']['PluginListResponse'];
type OpenApiPluginLifecycleRequest = components['schemas']['PluginLifecycleRequest'];
type OpenApiPluginLifecycleResponse = components['schemas']['PluginLifecycleResponse'];
type OpenApiPluginSettingsResponse = components['schemas']['PluginSettingsResponse'];
type OpenApiPluginSettingsUpdateRequest = components['schemas']['PluginSettingsUpdateRequest'];
type OpenApiPluginSettingsUpdateResponse = components['schemas']['PluginSettingsUpdateResponse'];

const _infoAnchor: keyof OpenApiPluginInfo = 'settings_count';
const _listAnchor: keyof OpenApiPluginListResponse = 'total';
const _lifecycleRequestAnchor: keyof OpenApiPluginLifecycleRequest = 'action';
const _lifecycleResponseAnchor: keyof OpenApiPluginLifecycleResponse = 'restart_required';
const _settingsAnchor: keyof OpenApiPluginSettingsResponse = 'masked_keys';
const _settingsRequestAnchor: keyof OpenApiPluginSettingsUpdateRequest = 'mask_token';
const _settingsUpdateAnchor: keyof OpenApiPluginSettingsUpdateResponse = 'changed_keys';
void _infoAnchor;
void _listAnchor;
void _lifecycleRequestAnchor;
void _lifecycleResponseAnchor;
void _settingsAnchor;
void _settingsRequestAnchor;
void _settingsUpdateAnchor;

export type PluginLifecycleAction = OpenApiPluginLifecycleRequest['action'];
export type PluginLifecycleState = OpenApiPluginInfo['state'];
export type PluginSource = OpenApiPluginInfo['source'];
export type PluginSettingValue = string | number | boolean | null;

export type PluginInfo = {
  id: string;
  name: string;
  version: string;
  source: PluginSource;
  state: PluginLifecycleState;
  desiredEnabled: boolean;
  reloadable: boolean;
  packageRoot?: string | null;
  extensionPoints: string[];
  lastErrorCode?: string | null;
  description: string;
  author: string;
  settingsCount: number;
};

export type PluginListResponse = {
  items: PluginInfo[];
  total: number;
};

export type PluginLifecycleResponse = {
  pluginId: string;
  action: PluginLifecycleAction;
  success: boolean;
  state: PluginLifecycleState;
  reloaded: boolean;
  restartRequired: boolean;
  errorCode?: string | null;
  message?: string | null;
  plugin?: PluginInfo | null;
};

export type PluginSettingField = {
  key: string;
  title: string;
  description: string;
  dataType: 'string' | 'integer' | 'number' | 'boolean';
  uiControl: 'text' | 'password' | 'number' | 'select' | 'textarea' | 'switch';
  isSensitive: boolean;
  isRequired: boolean;
  defaultValue: PluginSettingValue;
  options: Array<{ label: string; value: PluginSettingValue }>;
  validation: Record<string, unknown>;
  displayOrder: number;
};

export type PluginSettingsResponse = {
  pluginId: string;
  schema: PluginSettingField[];
  values: Record<string, PluginSettingValue>;
  maskedKeys: string[];
  maskToken: string;
};

export type PluginSettingsUpdateResponse = PluginSettingsResponse & {
  changedKeys: string[];
  restartRequired: boolean;
};

const finiteNumberSchema = z.number().finite();
const settingValueSchema = z.union([
  z.string(),
  z.number().int().finite(),
  finiteNumberSchema,
  z.boolean(),
  z.null(),
]);

const pluginInfoSchema = z.object({
  id: z.string(),
  name: z.string(),
  version: z.string(),
  source: z.enum(['builtin', 'external']),
  state: z.enum(['registered', 'enabled', 'disabled', 'failed']),
  desiredEnabled: z.boolean(),
  reloadable: z.boolean(),
  packageRoot: z.string().nullable().optional(),
  extensionPoints: z.array(z.string()).optional(),
  lastErrorCode: z.string().nullable().optional(),
  description: z.string().optional(),
  author: z.string().optional(),
  settingsCount: z.number().int().nonnegative().optional(),
}).passthrough();

const pluginListResponseSchema = z.object({
  items: z.array(pluginInfoSchema).optional(),
  total: z.number().int().nonnegative(),
}).passthrough();

const pluginLifecycleResponseSchema = z.object({
  pluginId: z.string(),
  action: z.enum(['enable', 'disable', 'reload']),
  success: z.boolean(),
  state: z.enum(['registered', 'enabled', 'disabled', 'failed']),
  reloaded: z.boolean().optional(),
  restartRequired: z.boolean().optional(),
  errorCode: z.string().nullable().optional(),
  message: z.string().nullable().optional(),
  plugin: pluginInfoSchema.nullable().optional(),
}).passthrough();

const pluginSettingFieldSchema = z.object({
  key: z.string(),
  title: z.string(),
  description: z.string().optional(),
  dataType: z.enum(['string', 'integer', 'number', 'boolean']),
  uiControl: z.enum(['text', 'password', 'number', 'select', 'textarea', 'switch']),
  isSensitive: z.boolean().optional(),
  isRequired: z.boolean().optional(),
  defaultValue: settingValueSchema.optional(),
  options: z.array(z.object({
    label: z.string(),
    value: settingValueSchema,
  })).optional(),
  validation: z.record(z.string(), z.unknown()).optional(),
  displayOrder: z.number().int().nonnegative().optional(),
}).passthrough();

const pluginSettingsResponseSchema = z.object({
  pluginId: z.string(),
  schema: z.array(pluginSettingFieldSchema).optional(),
  values: z.record(z.string(), settingValueSchema).optional(),
  maskedKeys: z.array(z.string()).optional(),
  maskToken: z.string(),
}).passthrough();

const pluginSettingsUpdateResponseSchema = pluginSettingsResponseSchema.extend({
  changedKeys: z.array(z.string()).optional(),
  restartRequired: z.boolean().optional(),
});

function normalizePlugin(raw: z.infer<typeof pluginInfoSchema>): PluginInfo {
  return {
    id: raw.id,
    name: raw.name,
    version: raw.version,
    source: raw.source,
    state: raw.state,
    desiredEnabled: raw.desiredEnabled,
    reloadable: raw.reloadable,
    packageRoot: raw.packageRoot ?? null,
    extensionPoints: raw.extensionPoints ?? [],
    lastErrorCode: raw.lastErrorCode ?? null,
    description: raw.description ?? '',
    author: raw.author ?? '',
    settingsCount: raw.settingsCount ?? 0,
  };
}

function normalizeSettings(
  raw: z.infer<typeof pluginSettingsResponseSchema>,
): PluginSettingsResponse {
  return {
    pluginId: raw.pluginId,
    schema: (raw.schema ?? []).map((field) => ({
      key: field.key,
      title: field.title,
      description: field.description ?? '',
      dataType: field.dataType,
      uiControl: field.uiControl,
      isSensitive: field.isSensitive ?? false,
      isRequired: field.isRequired ?? false,
      defaultValue: field.defaultValue ?? null,
      options: field.options ?? [],
      validation: field.validation ?? {},
      displayOrder: field.displayOrder ?? 100,
    })),
    values: raw.values ?? {},
    maskedKeys: raw.maskedKeys ?? [],
    maskToken: raw.maskToken,
  };
}

export const pluginsApi = {
  async list(): Promise<PluginListResponse> {
    const response = await apiClient.get<Record<string, unknown>>(
      '/api/v1/plugins',
      locallyRecoverableResourceConfig(),
    );
    const parsed = parseCamelCasePayload<z.infer<typeof pluginListResponseSchema>>(
      response.data,
      pluginListResponseSchema,
      'PluginListResponse',
      'plugins',
    );
    return { items: (parsed.items ?? []).map(normalizePlugin), total: parsed.total };
  },

  async updateLifecycle(
    pluginId: string,
    action: PluginLifecycleAction,
  ): Promise<PluginLifecycleResponse> {
    const response = await apiClient.post<Record<string, unknown>>(
      `/api/v1/plugins/${encodeURIComponent(pluginId)}/lifecycle`,
      { action } satisfies OpenApiPluginLifecycleRequest,
      locallyRecoverableResourceConfig(),
    );
    const parsed = parseCamelCasePayload<z.infer<typeof pluginLifecycleResponseSchema>>(
      response.data,
      pluginLifecycleResponseSchema,
      'PluginLifecycleResponse',
      'plugins',
    );
    return {
      pluginId: parsed.pluginId,
      action: parsed.action,
      success: parsed.success,
      state: parsed.state,
      reloaded: parsed.reloaded ?? false,
      restartRequired: parsed.restartRequired ?? false,
      errorCode: parsed.errorCode ?? null,
      message: parsed.message ?? null,
      plugin: parsed.plugin ? normalizePlugin(parsed.plugin) : null,
    };
  },

  async getSettings(pluginId: string): Promise<PluginSettingsResponse> {
    const response = await apiClient.get<Record<string, unknown>>(
      `/api/v1/plugins/${encodeURIComponent(pluginId)}/settings`,
      locallyRecoverableResourceConfig(),
    );
    const parsed = parseCamelCasePayload<z.infer<typeof pluginSettingsResponseSchema>>(
      response.data,
      pluginSettingsResponseSchema,
      'PluginSettingsResponse',
      'plugins',
    );
    return normalizeSettings(parsed);
  },

  async updateSettings(
    pluginId: string,
    values: Record<string, PluginSettingValue>,
    maskToken: string,
  ): Promise<PluginSettingsUpdateResponse> {
    const payload = { values, mask_token: maskToken } satisfies OpenApiPluginSettingsUpdateRequest;
    const response = await apiClient.put<Record<string, unknown>>(
      `/api/v1/plugins/${encodeURIComponent(pluginId)}/settings`,
      payload,
      locallyRecoverableResourceConfig(),
    );
    const parsed = parseCamelCasePayload<z.infer<typeof pluginSettingsUpdateResponseSchema>>(
      response.data,
      pluginSettingsUpdateResponseSchema,
      'PluginSettingsUpdateResponse',
      'plugins',
    );
    return {
      ...normalizeSettings(parsed),
      changedKeys: parsed.changedKeys ?? [],
      restartRequired: parsed.restartRequired ?? false,
    };
  },
};

File: src/api/v1/schemas/plugins.py
# Copyright (c) 2026 SiinXu / StockPulse contributors
# SPDX-License-Identifier: AGPL-3.0-only
"""Plugin lifecycle control API schemas (list + toggle/reload)."""

from __future__ import annotations

from typing import Dict, List, Literal, Optional, Union

from pydantic import BaseModel, ConfigDict, Field


PluginLifecycleAction = Literal["enable", "disable", "reload"]
PluginLifecycleState = Literal["registered", "enabled", "disabled", "failed"]
PluginSource = Literal["builtin", "external"]
PluginSettingValue = Union[str, int, float, bool, None]


class PluginInfo(BaseModel):
    """One registered plugin and its lifecycle state."""

    model_config = ConfigDict(extra="forbid")

    id: str = Field(..., description="Stable plugin id from the manifest")
    name: str
    version: str
    source: PluginSource
    state: PluginLifecycleState
    desired_enabled: bool = Field(
        ...,
        description="Persisted operator intent (False means disabled across restarts)",
    )
    reloadable: bool = Field(
        ...,
        description="True when in-process hot-reload is attempted for this plugin",
    )
    package_root: Optional[str] = Field(
        None,
        description="Absolute package directory for external plugins when known",
    )
    extension_points: List[str] = Field(default_factory=list)
    notification_channels: List[str] = Field(
        default_factory=list,
        description=(
            "Canonical notification channel IDs from active notification_channel "
            "registrations; empty when the adapter is not loaded or not active"
        ),
    )
    description: str = ""
    author: str = ""
    last_error_code: Optional[str] = Field(
        None,
        description="Stable last lifecycle failure code when the plugin is degraded",
    )
    settings_count: int = Field(
        0,
        ge=0,
        description="Number of declarative settings fields in the plugin manifest",
    )


class PluginListResponse(BaseModel):
    """GET /api/v1/plugins response."""

    model_config = ConfigDict(extra="forbid")

    items: List[PluginInfo] = Field(default_factory=list)
    total: int = Field(..., ge=0)


class PluginHealthEntryResponse(BaseModel):
    """One plugin health row for operator diagnostics consumers."""

    model_config = ConfigDict(extra="forbid")

    plugin_id: str
    name: str
    version: str
    source: PluginSource
    state: PluginLifecycleState
    desired_enabled: bool
    extension_points: List[str] = Field(default_factory=list)
    last_error_code: Optional[str] = None
    package_root: Optional[str] = None
    reloadable: bool = False


class PluginHealthResponse(BaseModel):
    """GET /api/v1/plugins/health — read-only plugin health snapshot."""

    model_config = ConfigDict(extra="forbid")

    generated_at: str = Field(
        ...,
        description="UTC ISO-8601 timestamp when the snapshot was built",
    )
    total: int = Field(..., ge=0)
    plugins: List[PluginHealthEntryResponse] = Field(default_factory=list)


class PluginLifecycleRequest(BaseModel):
    """POST /api/v1/plugins/{plugin_id}/lifecycle body."""

    model_config = ConfigDict(extra="forbid")

    action: PluginLifecycleAction = Field(
        ...,
        description="enable/disable toggles runtime state and persistence; reload re-imports code",
    )


class PluginLifecycleResponse(BaseModel):
    """Result of one lifecycle action."""

    model_config = ConfigDict(extra="forbid")

    plugin_id: str
    action: PluginLifecycleAction
    success: bool
    state: PluginLifecycleState
    reloaded: bool = False
    restart_required: bool = False
    error_code: Optional[str] = None
    message: Optional[str] = None
    plugin: Optional[PluginInfo] = None


class PluginSettingOptionResponse(BaseModel):
    """One finite option for a generated plugin setting control."""

    model_config = ConfigDict(extra="forbid", strict=True, allow_inf_nan=False)

    label: str
    value: PluginSettingValue


class PluginSettingFieldResponse(BaseModel):
    """One strict manifest-declared plugin settings field."""

    model_config = ConfigDict(extra="forbid", strict=True, allow_inf_nan=False)

    key: str
    title: str
    description: str = ""
    data_type: Literal["string", "integer", "number", "boolean"]
    ui_control: Literal["text", "password", "number", "select", "textarea", "switch"]
    is_sensitive: bool = False
    is_required: bool = False
    default_value: PluginSettingValue = None
    options: List[PluginSettingOptionResponse] = Field(default_factory=list)
    validation: Dict[str, object] = Field(default_factory=dict)
    display_order: int = Field(100, ge=0)


class PluginSettingsResponse(BaseModel):
    """Generated schema and masked effective values for one plugin."""

    model_config = ConfigDict(extra="forbid", strict=True, allow_inf_nan=False)

    plugin_id: str
    schema_: List[PluginSettingFieldResponse] = Field(default_factory=list, alias="schema")
    values: Dict[str, PluginSettingValue] = Field(default_factory=dict)
    masked_keys: List[str] = Field(default_factory=list)
    mask_token: str = "******"


class PluginSettingsUpdateRequest(BaseModel):
    """Full replacement of explicit values; omitted keys reset to defaults."""

    model_config = ConfigDict(extra="forbid", strict=True, allow_inf_nan=False)

    values: Dict[str, PluginSettingValue] = Field(default_factory=dict)
    mask_token: str = "******"


class PluginSettingsUpdateResponse(PluginSettingsResponse):
    """Persisted settings projection plus apply semantics."""

    changed_keys: List[str] = Field(default_factory=list)
    restart_required: bool = False
